aiida/schedulers/scheduler.py

```diff
@@ -216,7 +216,7 @@ def _get_run_line(self, codes_info, codes_run_mode):
 
             output_string = f'{command_to_exec} {stdin_str} {stdout_str} {stderr_str}'
 
-            list_of_runlines.append(output_string)
+            list_of_runlines.append(output_string.rstrip())
```

Comment:
Hey @unkcpz why did you change this here? The `rstrip()` on the test when you are doing the `file_regression` should suffice for the test to work, no? I wouldn't change this in the actual code just to cater for a test or comply with some linter criteria in a script that is produced during usage and not part of the codebase. I mean, this is changing some intermediate file produce by the code that could be checked or parsed by some users, and its modification may create some problems (however unlikely it might be in this case).

I mean, unless you have a strong reason to propose we change to having our scripts be trimmed at the end, I would say to revert this change.

Reply:
This is intended to pass the linter. As you can see from the above line, that the `output_string` will end will spaces if `stdin_str`, `stdout_str` or `stderr_str` is not set. In that case, the command line of submit script will end with spaces wiped out by the linter and fails the pre-commit. 

> I mean, this is changing some intermediate file produce by the code that could be checked or parsed by some users, and its modification may create some problems (however unlikely it might be in this case

I agree, but for this part of the code, I see no downside to strip the line. In other words, I would much more like the bash script ironed without extra spaces in each line ;-)

Reply:
> This is intended to pass the linter. As you can see from the above line, that the `output_string` will end will spaces if `stdin_str`, `stdout_str` or `stderr_str` is not set. In that case, the command line of submit script will end with spaces wiped out by the linter and fails the pre-commit.

Right, but I mean, you are not committing any script that is being generated by this, right? The reference file is being created by `file_regression.check(content, extension='.sh')` and you are already using `rstrip` to clean up `content` before putting it there. That you originally take `content` from the printed script does not matter if you clean it between reading it and writing to reference with the `file_regression`. Am I missing something here? Which script directly generated by this is being rejected by the linter?

> I agree, but for this part of the code, I see no downside to strip the line. In other words, I would much more like the bash script ironed without extra spaces in each line ;-)

I also agree that it is very unlikely that this will break someone's code, but it is still possible. If we have had a team discussion on how to do a cleaner script and agreed in some guidelines that we were going to respect going forward, I would totally bite the bullet on the chance of causing some friction. But we are not even sure that we won't have to change this further (maybe to prevent contiguous empty lines due to non existing append/prepend texts) or even go back on it (maybe it turns out there are good reasons to have one final empty line at the end of the file). So yeah, the risk may be minimal but on the other hand there is no objective gain from this change (other than you and me liking it more aesthetically, but that is not very objective =P).

Reply:
Sorry for not explain it more clear. 

I have to add this to pass the linter. Because the `content.rstrip()` in test function, only removes the final lines of the submit script content. `output_string` is the executable command line not necessary the final line. For clarity, please see the following content generated by in test and after linter pre-commit clean up (I use '␣' to represent space character):

The script generated:
```bash
#!/bin/bash
exec > _scheduler-stdout.txt
exec 2> _scheduler-stderr.txt


'/bin/bash'␣␣␣

'mpirun' '-np' '1' '/bin/bash'
```

After pre-commit:
```bash
#!/bin/bash
exec > _scheduler-stdout.txt
exec 2> _scheduler-stderr.txt


'/bin/bash'

'mpirun' '-np' '1' '/bin/bash'
```

If we agree not to add exclude files in the linter configuration, I can not figure out a simple way to get rid of this and pass the pre-commit check.

Reply:
You are absolutely right @unkcpz the formatter strips all whitespace, also terminating spaces in lines in the middle of the file. Apologies for not having thought of this, I was focusing on the terminating new line characters.

OK, so here is my thought process. We have added the whitespace trimming pre-commit action such that trailing whitespace is automatically cleaned up in source code, as it can make things look messy and more difficult to read than necessary. When I added it, I did not specify what files it should run for, and so let it run for everything, even comparison files that are used in the unit tests, even though those are not the primary target. We don't need to read those often, so trailing newspace is not a real problem. When you consider the motivation for the whitespace trimming, it seems clear that the correct solution for this problem is to exclude the test comparison file from the pre-commit. So I now agree that @unkcpz original solution of updating the exclude rules in the `.pre-commit-config.yaml` file is actually the correct one.

I propose therefore that we revert the changes we requested (really sorry for the back and forth @unkcpz ), removing the manual explicit whitespace stripping in the code and the test and simply update the include rules in the config. Note that instead of excluding explicitly this file, maybe it would be better to simply _include_ all `.py` files, because that is in the end what we are really interested in. We want to clean the source code files, not any other arbitrary text files.

What we could do is use the following regex I think
```
        exclude: >- >
            (?x)^(
                tests/.*(?<!\.py)$
            )$
```

i.e. exclude all files in the `tests` folder that _do not_ finish with the `.py` extension. We still want to clean the test source files of course. Would you agree with this suggestion @ramirezfranciscof ?

Reply:
@unkcpz Ahhhh I see now! Thanks for the explanation, haha, indeed I can see the excess with the newlines but it is more difficult to notice the trailing spaces 😅. Nice idea using the strange underscore symbol to illustrate it.

@sphuber mmm I'm trying to think if this could backfire. I mean, the test section does have other kinds of file (the `test/static` folders have some `.cif`, `.UPF`, `.aiida` exports, `.json`) but perhaps we don't want to be lintering those anyways? I also had the idea that there were some notebooks being used to do some tests, but I couldn't find anything there now. I'm wondering if at some point it won't be convenient to use some bash scripts as aux to some tests and we may want to linter those too. It is also worth mentioning that the `.sh` extension was set explicitly by `file_regression.check(content, extension='.sh')`, so we could also decide on a specific extension for reference files and exclude those (`.pyreg`? `.pytest_regression_uniquely_long_aiida_extension`?). I mean, I agree it would be nice to set some configuration to deal now and for all with all pytest regression files, I'm just not sure just applying the linter to `.py` is the clear way to do this.

There is also one last possible chance that we might be able to deal with this in a simple way inside the test (which is, actually, more self-documenting that the initial one I proposed...):

``` python
    content = '# pylint: skip-file\n' + content
```

In any case, as the proposed alternative would be a modification internal to the tests so we are more free to set it back if we change our minds, I'm ok with you going the way you prefer. It still pains me to have to add the exception to the `.pre-commit-config.yaml`, but since you already tried a suggestion I made that underestimated the problem I won't insist on this if you think the pylint config is the best way to go.

Reply:
> There is also one last possible chance that we might be able to deal with this in a simple way inside the test (which is, actually, more self-documenting that the initial one I proposed...):
> 
> ```python
>     content = '# pylint: skip-file\n' + content
> ```

It is not `pylint` that does the trimming though, so that won't work.

Reply:
Mmm I see. I can't find anywhere how to signal inside the file that no hooks should be applied to it, so perhaps this is not possible in a general way. Weird. Anyways, after two unsuccessful suggestions I concede, feel free to go ahead with @sphuber suggested solution then @unkcpz =)